libraries/rust/data/src/client.rs

```diff
@@ -0,0 +1,218 @@
+        let mut equities_bars: Vec<Bar> = Vec::new();
+
+        let mut headers = HeaderMap::new();
+        headers.insert(ACCEPT, HeaderValue::from_str("application/json")?);
+        headers.insert("APCA-API-KEY-ID", HeaderValue::from_str(&self.alpaca_api_key_id)?);
+        headers.insert("APCA-API-SECRET-KEY", HeaderValue::from_str(&self.alpaca_api_secret_key)?);
+
+        for ticker in tickers {
+            loop {
+                let mut page_token: Option<String> = None;
+
+                let alpaca_url = self.build_equities_bars_url(&ticker, start, end, page_token.as_deref())?;
+
+                let response = self.http_client.get(alpaca_url)
+                    .headers(headers.clone())
+                    .send()
+                    .await?;
+        
+                if !response.status().is_success() {
+                    return Err(format!("api request failed with status: {}", response.status()).into());
+                }
+
+                let bar_response: BarsResponse = response.json().await?;
+        
+                let bars_with_ticker = bar_response.bars.into_iter()
+                    .map(|mut bar| {
+                        bar.ticker = Some(ticker.clone());
+                        bar
+                    })
+                    .collect::<Vec<Bar>>();
+        
+                equities_bars.extend(bars_with_ticker);
+        
+                match bar_response.next_token {
+                    Some(token) => page_token = Some(token),
+                    None => break,
+                }
+            }
+        }
+
+        Ok(equities_bars)
+    }
+
+    fn build_equities_bars_url(
+        &self,
+        ticker: &String,
+        start: DateTime<Utc>,
+        end: DateTime<Utc>,
+        page_token: Option<&str>,
+    ) -> Result<Url, Box<dyn std::error::Error>> {
+        let path = format!("stocks/{}/bars", ticker);
+
+        let start_str = start.format("%Y-%m-%dT%H:%M:%SZ").to_string();
+        let end_str = end.format("%Y-%m-%dT%H:%M:%SZ").to_string();
+
+        let mut alpaca_url = Url::parse(&self.alpaca_base_url)?
+            .join(&path)?;
+
+        {
+            let mut query_pairs = alpaca_url.query_pairs_mut();
+            
+            query_pairs.append_pair("timeframe", "1D")
+                .append_pair("start", &start_str)
+                .append_pair("end", &end_str)
+                .append_pair("limit", "10000")
+                .append_pair("adjustment", "all")
+                .append_pair("feed", "sip")
+                .append_pair("sort", "asc");
+
+            if let Some(token) = page_token {
+                query_pairs.append_pair("page_token", &token);
+            }
+        }
+
+        Ok(alpaca_url)
+    }
+
+    pub async fn write_equities_bars_to_storage(&self, equities_bars: Vec<Bar> ) -> Result<(), Box<dyn std::error::Error>> {
+        let equities_bars_json = serde_json::to_vec(&equities_bars).unwrap();
```

Comment:
we should remove `unwrap` when possible and handle explicitly `?`

https://rust-lang.github.io/rust-clippy/master/#/unwrap_used